service/history/api/get_workflow_util.go

```diff
@@ -234,6 +248,55 @@ func GetMutableState(
 	return MutableStateToGetResponse(mutableState)
 }
 
+func GetMutableStateWithConsistencyCheck(
+	ctx context.Context,
+	shardContext shard.Context,
+	workflowKey definition.WorkflowKey,
+	currentVersion int64,
+	currentEventID int64,
+	workflowConsistencyChecker WorkflowConsistencyChecker,
+) (_ *historyservice.GetMutableStateResponse, retError error) {
+
+	if len(workflowKey.RunID) == 0 {
+		return nil, serviceerror.NewInternal(fmt.Sprintf(
+			"getMutableState encountered empty run ID: %v", workflowKey,
+		))
+	}
+
+	workflowLease, err := workflowConsistencyChecker.GetWorkflowLeaseWithConsistencyCheck(
+		ctx,
+		nil,
+		func(mutableState workflow.MutableState) bool {
+			mutableState.GetExecutionInfo().GetVersionHistories()
+			currentVersionHistory, err := versionhistory.GetCurrentVersionHistory(mutableState.GetExecutionInfo().GetVersionHistories())
+			if err != nil {
+				return false
+			}
+			lastVersionHistoryItem, err := versionhistory.GetLastVersionHistoryItem(currentVersionHistory)
+			if err != nil {
+				return false
+			}
+
+			if currentVersion == lastVersionHistoryItem.GetVersion() {
+				return currentEventID <= lastVersionHistoryItem.GetEventId()
+			}
+			return currentVersion < lastVersionHistoryItem.GetVersion()
```

Comment:
oh I realized with state-based replication, we need to change token to use versioned transition... In state-based replication world, we do allow event version version to go backwards as long as the overall versioned transition is keep increasing. I need to track this.

Reply:
I see. I think it needs to compare the versioned transition first.

Reply:
I will add a TODO for now.